parser/shared/src/test/scala/hmda/parser/fi/lar/LarGenerators.scala

```diff
@@ -6,30 +6,125 @@ import org.scalacheck.Gen
-
-  implicit def actionDateGen: Gen[Int] = ???
-
-  implicit def geographyGen: Gen[Geography] = ???
-
-  implicit def applicantGen: Gen[Applicant] = ???
-
-  implicit def purchaserTypeGen: Gen[Int] = ???
-
-  implicit def denialGen: Gen[Denial] = ???
-
-  implicit def rateSpreadGen: Gen[String] = ???
-
-  implicit def hoepaStatusGen: Gen[Int] = ???
-
-  implicit def lienStatusGen: Gen[Int] = ???
-
+  implicit def larGen: Gen[LoanApplicationRegister] = {
+    for {
+      respondent <- respIdGen
+      agencyCode <- agencyCodeGen
+      loan <- loanGen
+      preapprovals <- preapprovalGen
+      actionTaken <- actionTypeGen
+      actionTakenDate <- actionDateGen
+      geography <- geographyGen
+      applicant <- applicantGen
+      purchaserType <- purchaserTypeGen
+      denial <- denialGen
+      rateSpread <- rateSpreadGen
+      hoepaStatus <- hoepaStatusGen
+      lienStatus <- lienStatusGen
+    } yield LoanApplicationRegister(
+      2, // TODO we may not want to hard-code this everywhere (though it may be appropriate here)
+      respondent,
+      agencyCode,
+      loan,
+      preapprovals,
+      actionTaken,
+      actionTakenDate,
+      geography,
+      applicant,
+      purchaserType,
+      denial,
+      rateSpread,
+      hoepaStatus,
+      lienStatus
+    )
+  }
+
+  implicit def loanGen: Gen[Loan] = {
+    for {
+      id <- Gen.listOf(Gen.alphaNumChar).map(_.mkString)
+      //applicationDate <- ??? // TODO Format is ccyymmdd or NA
+      loanType <- Gen.oneOf(1, 2, 3, 4)
+      propertyType <- Gen.oneOf(1, 2, 3)
+      purpose <- Gen.oneOf(1, 2, 3)
+      occupancy <- Gen.oneOf(1, 2, 3)
+      amount <- Gen.posNum[Int] // TODO is that right? can it be zero? (what would a loan for zero thousand dollars mean?)
+    } yield Loan(
+      id,
+      "NA",
+      loanType,
+      propertyType,
+      purpose,
+      occupancy,
+      amount
+    )
+  }
+
+  implicit def preapprovalGen: Gen[Int] = Gen.oneOf(1, 2, 3)
+
+  implicit def actionTypeGen: Gen[Int] = Gen.choose(1, 8)
+
+  implicit def actionDateGen: Gen[Int] = Gen.choose(20170101, 20201231) // TODO this allows non-date numbers!
+  // so play with things like LocalDate.of(2017, 01, 01) and figure out how to convert
+
+  implicit def geographyGen: Gen[Geography] = { // TODO any of these can be NA
+    for {
+      msa <- stringOfN(5, Gen.numChar) // actually it's more specific; do we care, for this purpose?
+      state <- stringOfN(2, Gen.numChar) // TODO limit to real ones? http://www2.census.gov/geo/docs/reference/state.txt
+      county <- stringOfN(3, Gen.numChar)
+      tract <- censusTractGen
+    } yield Geography(
+      msa,
+      state,
+      county,
+      tract
+    )
+  }
+
+  implicit def censusTractGen = {
+    for {
+      tract <- stringOfN(4, Gen.numChar).map(Some(_)) // TODO are leading zeroes required? (they are allowed.)
+      suffix <- Gen.option(stringOfN(2, Gen.numChar))
+    } yield List(tract, Some("."), suffix).flatten.mkString
+  }
```

Comment:
I believe leading zeroes are required. Also, If I'm reading this one correctly it can generate empty suffix portion, I think that's not valid